pyk/src/pyk/kast.py

```diff
@@ -349,23 +371,50 @@ def map_inner(self: 'KLabel', f: Callable[[KInner], KInner]) -> 'KLabel':
     def match(self, term: KInner) -> Optional[Subst]:
         raise TypeError('KLabel does not support pattern matching')
 
+    @overload
+    def apply(self, args: Iterable[KInner]) -> 'KApply':
+        ...
+
+    @overload
+    def apply(self, *args: KInner) -> 'KApply':
+        ...
+
+    def apply(self, *args, **kwargs):
+        return KApply(self, *args, **kwargs)
+
 
 @final
 @dataclass(frozen=True)
 class KApply(KInner):
     label: KLabel
     args: Tuple[KInner, ...]
 
-    def __init__(self, label: Union[str, KLabel], args: Iterable[KInner] = ()):
+    @overload
+    def __init__(self, label: Union[str, KLabel], args: Iterable[KInner]):
+        ...
+
+    @overload
+    def __init__(self, label: Union[str, KLabel], *args: KInner):
+        ...
+
+    def __init__(self, label, *args, **kwargs):
         if type(label) is str:
             label = KLabel(label)
 
+        if kwargs:
+            bad_arg = next((arg for arg in kwargs if arg != 'args'), None)
+            if bad_arg:
+                raise TypeError(f"KApply() got an unexpected keyword argument '{bad_arg}'")
+            if args:
+                raise TypeError("KApply() got multiple values for argument 'args'")
+            args = kwargs['args']
+
+        elif len(args) == 1 and isinstance(args[0], Iterable) and not isinstance(args[0], KInner):
+            args = args[0]
+
         object.__setattr__(self, 'label', label)
```

Comment:
I'm really having trouble trying to quickly understand what functionality this gives us. Can you add a small comment explaining the use case of this constructor?

Reply:
Also, how does constructor overloading work here? How will python know which `__init__` to call?

Reply:
There is a single method with actual implementation, that's the one that's going to be invoked during execution. Declarations with the `@overload` decorator only provide type hints for the type checker.

In this case, the `@overload` declarations express that `KApply(f, x, y, z)` and `KApply(f, [x, y, z])` should type check, and the implementation makes sure that these invocations are handled correctly (I added tests to check this).